### PR TITLE
feat: add mutex for pkam and cram authentication methods

### LINKED ISSUE #203: Race possible while authenticating
**Describe the bug**
If the `authenticate(String? privateKey)` in AtLookupImpl is called twice in quick succession, then the requests and responses for the two could become intermingled, whereas the expectation is that the entire authentication flow should complete atomically

**Expected behavior**
1. Employ a mutex, similar to how the `_process` method works
2. Within the critical section of code, add a check to begin with of the boolean which tells us whether the connection has already been authenticated should be

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
1. Add mutex around authenticate and authenticate_cram methods.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 1]
```diff
@@ -350,63 +350,77 @@ class AtLookupImpl implements AtLookUp {
...
-    await _sendCommand('pkam:$signature\n');
-    var pkamResponse = await messageListener.read();
-    if (pkamResponse == 'data:success') {
-      logger.info('auth success');
-      _isPkamAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException(
-          'Failed connecting to $_currentAtSign. $pkamResponse');
+    try {
+      pkamAuthenticationMutex.acquire();
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: Wrap the next block of code in an if condition
```
if (!_isPkamAuthenticated) {
    await _sendCommand ..... etc
}
```

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 2]
```diff
@@ -350,63 +350,77 @@ class AtLookupImpl implements AtLookUp {
...
-    var digest = sha512.convert(bytes);
-    await _sendCommand('cram:$digest\n');
-    var cramResponse = await messageListener.read();
-    if (cramResponse == 'data:success') {
-      logger.info('auth success');
-      _isCramAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException('Auth failed');
+    try {
+      cramAuthenticationMutex.acquire();
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: Wrap the next block of code in an if condition
```
if (!_isCramAuthenticated) {
    await _sendCommand ..... etc
}
```

**Review by gkc — CHANGES_REQUESTED** [review]

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: The next block code is moved into IF condition.

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: The next block of code is moved into IF condition.

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 3]
```diff
@@ -350,63 +350,85 @@ class AtLookupImpl implements AtLookUp {
...
+          _isPkamAuthenticated = true;
+        } else {
+          throw UnAuthenticatedException(
+              'Failed connecting to $_currentAtSign. $pkamResponse');
+        }
+      }
+      return _isPkamAuthenticated;
+    } finally {
+      if (_pkamAuthenticationMutex.isLocked) {
+        _pkamAuthenticationMutex.release();
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: if condition not required here

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 4]
```diff
@@ -350,63 +350,85 @@ class AtLookupImpl implements AtLookUp {
...
+          logger.info('auth success');
+          _isCramAuthenticated = true;
+        } else {
+          throw UnAuthenticatedException('Auth failed');
+        }
+      }
+      return _isCramAuthenticated;
+    } finally {
+      if (_cramAuthenticationMutex.isLocked) {
+        _cramAuthenticationMutex.release();
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: if condition not required here

**Review by gkc — CHANGES_REQUESTED** [review]

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 4 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: The functional test in at_onboarding_cli have failed with the following error. Hence added the IF condition to check if lock is applied or not.

```
Bad state: no lock to release
  package:mutex/src/read_write_mutex.dart 164:7                         ReadWriteMutex.release
  package:mutex/src/mutex.dart 54:30                                    Mutex.release
  package:at_lookup/src/at_lookup_impl.dart 389:31                      AtLookupImpl.authenticate
  package:at_onboarding_cli/src/at_onboarding_service_impl.dart 194:28  AtOnboardingServiceImpl.authenticate
  ===== asynchronous gap ===========================
  test/at_onboarding_cli_test.dart 128:22
```

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 5]
```diff
@@ -350,63 +350,85 @@ class AtLookupImpl implements AtLookUp {
...
-    if (pkamResponse == 'data:success') {
-      logger.info('auth success');
-      _isPkamAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException(
-          'Failed connecting to $_currentAtSign. $pkamResponse');
+    try {
+      if (!_isPkamAuthenticated) {
+        _pkamAuthenticationMutex.acquire();
+        await _sendCommand('from:$_currentAtSign\n');
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart:364]: Move `_pkamAuthenticationMutex.acquire();` to before the if statement.

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 6]
```diff
@@ -350,63 +350,85 @@ class AtLookupImpl implements AtLookUp {
...
-    await _sendCommand('cram:$digest\n');
-    var cramResponse = await messageListener.read();
-    if (cramResponse == 'data:success') {
-      logger.info('auth success');
-      _isCramAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException('Auth failed');
+    try {
+      if (!_isCramAuthenticated) {
+        _cramAuthenticationMutex.acquire();
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: Move `_cramAuthenticationMutex.acquire();` to before the if statement.

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 4 above)
Comment (gkc) [reply at_lookup/lib/src/at_lookup_impl.dart]: That's because I see you aren't acquiring the mutex every time. I've added comments above

**Review by gkc — CHANGES_REQUESTED** [review]

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 4 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: Got it, My bad Gary. I missed the below comments. Will make the necessary changes. Thank you.

**Review by gkc — APPROVED** [review]

**Pull request merged**